win32/src/win32pdhmodule.cpp

```diff
@@ -744,6 +754,74 @@ static PyObject *PyGetFormattedCounterValue(PyObject *self, PyObject *args)
 	return realrc;
 }
 
+// @pymethod dictionary|win32pdh|GetFormattedCounterArray|Retrieves an array of formatted counter values
+static PyObject *PyPdhGetFormattedCounterArray(PyObject *self, PyObject *args)
+{
+	HCOUNTER handle;
+	PyObject *obhandle;
+	DWORD format;
+	if (!PyArg_ParseTuple(args, "Oi:PdhGetFormattedCounterArray", 
+			&obhandle, // @pyparm int|handle||Handle to the counter
+			&format)) // @pyparm int|format||Format of result.  Can be PDH_FMT_DOUBLE, PDH_FMT_LARGE, PDH_FMT_LONG and or'd with PDH_FMT_NOSCALE, PDH_FMT_1000
+
+		return NULL;
+	if (!PyWinObject_AsHANDLE(obhandle, &handle))
+		return NULL;
+	DWORD type;
+	PDH_FMT_COUNTERVALUE result;
+	CHECK_PDH_PTR(pPdhGetFormattedCounterArray);
+	PDH_STATUS pdhStatus;
+	DWORD size = 0;
+	DWORD count;
+	PDH_FMT_COUNTERVALUE_ITEM *pItems = NULL;
+
+	Py_BEGIN_ALLOW_THREADS
+	pdhStatus = (*pPdhGetFormattedCounterArray) (handle, format, &size, &count, pItems);
+	Py_END_ALLOW_THREADS
+	if (pdhStatus != PDH_MORE_DATA)
+		return PyWin_SetAPIError("PdhGetFormattedCounterArray", pdhStatus);
+	pItems = (PDH_FMT_COUNTERVALUE_ITEM *) malloc(size);
+	if (pItems == NULL) {
+		PyErr_NoMemory();
+		return NULL;
+	}
+    
+	Py_BEGIN_ALLOW_THREADS
+	pdhStatus = (*pPdhGetFormattedCounterArray) (handle, format, &size, &count, pItems);
+	Py_END_ALLOW_THREADS
+	if (pdhStatus != ERROR_SUCCESS) {
+        free(pItems);
+		return PyWin_SetAPIError("PdhGetFormattedCounterArray", pdhStatus);
+	}
+
+	PyObject *key;
+	PyObject *value;
+	PyObject *rc;
+	rc = PyDict_New();
+	if (rc == NULL) {
+		free(pItems);
+		PyErr_NoMemory();
+		return NULL;
+	}
+	for (DWORD i = 0; i < count; i++) {
+		key = PyWinObject_FromTCHAR(pItems[i].szName);
+		if (format & PDH_FMT_DOUBLE)
+			value = PyFloat_FromDouble(pItems[i].FmtValue.doubleValue);
+		else if (format & PDH_FMT_LONG)
+			value = PyInt_FromLong(pItems[i].FmtValue.longValue);
+		else if (format & PDH_FMT_LARGE)
+			value = PyLong_FromLongLong(pItems[i].FmtValue.largeValue);
+		else {
+			PyErr_SetString(PyExc_ValueError, "Dont know how to convert the result");
```

Comment:
This seems wrong - you should probably DECREF `rc` and set it to null and break out of the loop, otherwise (a) you'll pass null to `PyDict_SetItem` and (b) python will make some noise about an error being set on non-error result.